CosNFileSystem.java

```diff
@@ -0,0 +1,814 @@
+
+    BufferPool.getInstance().initialize(this.getConf());
+
+    // initialize the thread pool
+    int uploadThreadPoolSize = this.getConf().getInt(
+        CosNConfigKeys.UPLOAD_THREAD_POOL_SIZE_KEY,
+        CosNConfigKeys.DEFAULT_UPLOAD_THREAD_POOL_SIZE
+    );
+    int readAheadPoolSize = this.getConf().getInt(
+        CosNConfigKeys.READ_AHEAD_QUEUE_SIZE,
+        CosNConfigKeys.DEFAULT_READ_AHEAD_QUEUE_SIZE
+    );
+    int ioThreadPoolSize = uploadThreadPoolSize + readAheadPoolSize / 3;
+    long threadKeepAlive = this.getConf().getLong(
+        CosNConfigKeys.THREAD_KEEP_ALIVE_TIME_KEY,
+        CosNConfigKeys.DEFAULT_THREAD_KEEP_ALIVE_TIME
+    );
+    this.boundedIOThreadPool = BlockingThreadPoolExecutorService.newInstance(
+        ioThreadPoolSize / 2, ioThreadPoolSize,
+        threadKeepAlive, TimeUnit.SECONDS,
+        "cos-transfer-thread-pool");
+    int copyThreadPoolSize = this.getConf().getInt(
+        CosNConfigKeys.COPY_THREAD_POOL_SIZE_KEY,
+        CosNConfigKeys.DEFAULT_COPY_THREAD_POOL_SIZE
+    );
+    this.boundedCopyThreadPool = BlockingThreadPoolExecutorService.newInstance(
+        CosNConfigKeys.DEFAULT_COPY_THREAD_POOL_SIZE, copyThreadPoolSize,
+        60L, TimeUnit.SECONDS,
+        "cos-copy-thread-pool");
+  }
+
+  private static NativeFileSystemStore createDefaultStore(Configuration conf) {
+    NativeFileSystemStore store = new CosNativeFileSystemStore();
+    RetryPolicy basePolicy = RetryPolicies.retryUpToMaximumCountWithFixedSleep(
+        conf.getInt(CosNConfigKeys.COSN_MAX_RETRIES_KEY,
+            CosNConfigKeys.DEFAULT_MAX_RETRIES),
+        conf.getLong(CosNConfigKeys.COSN_RETRY_INTERVAL_KEY,
+            CosNConfigKeys.DEFAULT_RETRY_INTERVAL),
+        TimeUnit.SECONDS);
+    Map<Class<? extends Exception>, RetryPolicy> exceptionToPolicyMap =
+        new HashMap<>();
+
+    exceptionToPolicyMap.put(IOException.class, basePolicy);
+    RetryPolicy methodPolicy = RetryPolicies.retryByException(
+        RetryPolicies.TRY_ONCE_THEN_FAIL,
+        exceptionToPolicyMap);
+    Map<String, RetryPolicy> methodNameToPolicyMap = new HashMap<>();
+    methodNameToPolicyMap.put("storeFile", methodPolicy);
+    methodNameToPolicyMap.put("rename", methodPolicy);
+
+    return (NativeFileSystemStore) RetryProxy.create(
+        NativeFileSystemStore.class, store, methodNameToPolicyMap);
+  }
+
+  private String getOwnerId() {
+    return System.getProperty("user.name");
+  }
+
+  private String getGroupId() {
+    return System.getProperty("user.name");
+  }
+
+  private String getOwnerInfo(boolean getOwnerId) {
+    String ownerInfoId = "";
+    try {
+      String userName = System.getProperty("user.name");
+      String command = "id -u " + userName;
+      if (!getOwnerId) {
+        command = "id -g " + userName;
+      }
+      Process child = Runtime.getRuntime().exec(command);
+      child.waitFor();
+
+      // Get the input stream and read from it
+      InputStream in = child.getInputStream();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *in*. The resource is referenced in statements at the following line: 188. The resource closure statement is at line: 191. There are other execution paths that do not contain closure statements, e.g., when exception is thrown by InputStream.read. Either a) close *in* in a try-finally block or b) close the resource by declaring *in* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).